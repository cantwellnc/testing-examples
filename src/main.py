import boto3
from mypy_boto3_s3 import S3Client
from mypy_boto3_s3.type_defs import (
    ListBucketsOutputTypeDef,
    GetBucketTaggingOutputTypeDef,
)

# Made up task: write a script that returns the # of buckets with PII
# (Personally Identifying Information) in an AWS account,
#  based on a 'pii' tag.


def my_list_bucket(client: S3Client) -> ListBucketsOutputTypeDef:
    return client.list_buckets()


def get_tags_for_all_buckets(
    buckets: list, client: S3Client
) -> list[GetBucketTaggingOutputTypeDef]:
    return [
        client.get_bucket_tagging(Bucket=bucket["Name"]) for bucket in buckets
    ]


def has_pii_content(bucket_tags: GetBucketTaggingOutputTypeDef) -> bool:
    # pii tag is of form {"Key": "pii", "Value": "true | false"}
    # intentionally bad implementation
    return bucket_tags["TagSet"][0]["Value"] == "true"
    # a better implementation
    # for tag in bucket_tags["TagSet"]:
    #     if tag["Key"] == "pii" and tag["Value"] == "true":
    #         return True
    # return False


def handler():
    # imagine this being the handler of a lambda that is running in your account
    # to document this info on a daily basis
    client = boto3.client("s3")
    buckets = my_list_bucket(client)
    tags = get_tags_for_all_buckets(buckets["Buckets"], client)
    return sum([has_pii_content(tag) for tag in tags])
