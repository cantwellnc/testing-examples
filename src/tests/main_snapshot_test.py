import pytest
from unittest.mock import MagicMock
from main import (
    get_tags_for_all_buckets,
    handler,
    has_pii_content,
    my_list_bucket,
)
from syrupy.assertion import SnapshotAssertion


@pytest.fixture
def mock_s3():
    mock = MagicMock()
    mock.list_bucket.return_value = {
        "Buckets": [
            {"Name": f"string{i}", "CreationDate": "2015-09-14T20:00:00.000Z"}
            for i in range(6)
        ],
        "Owner": {"DisplayName": "string", "ID": "string"},
    }

    def get_bucket_tagging_responses(Bucket: str, *args, **kwargs):
        if "5" in Bucket:
            return {"TagSet": [{"Key": "pii", "Value": "true"}]}
        return {"TagSet": [{"Key": "pii", "Value": "false"}]}

    mock.get_bucket_tagging.side_effect = get_bucket_tagging_responses
    return mock


@pytest.fixture
def positive_pii_tags():
    return {"TagSet": [{"Key": "pii", "Value": "true"}]}


@pytest.fixture
def negative_pii_tags():
    return {"TagSet": [{"Key": "pii", "Value": "false"}]}


def test_list_bucket(snapshot: SnapshotAssertion, mock_s3):
    assert snapshot == my_list_bucket(mock_s3)


def test_get_tags_for_all_buckets(snapshot: SnapshotAssertion, mock_s3):
    buckets = my_list_bucket(mock_s3)["Buckets"]
    result = get_tags_for_all_buckets(buckets, mock_s3)
    assert snapshot == result


def test_has_pii_content(
    snapshot: SnapshotAssertion, positive_pii_tags, negative_pii_tags
):
    # not really useful to test w snapshot since the output data structure is so simple
    assert has_pii_content(positive_pii_tags) == snapshot
    assert has_pii_content(negative_pii_tags) == snapshot


def test_handler(snapshot: SnapshotAssertion, mock_s3, mocker):
    mocker.patch("main.boto3.client", return_value=mock_s3)
    result = handler()
    assert result == snapshot  # expected bc of how we build mock_s3
