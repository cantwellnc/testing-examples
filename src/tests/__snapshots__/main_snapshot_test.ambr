# serializer version: 1
# name: test_get_tags_for_all_buckets
  list([
  ])
# ---
# name: test_handler
  0
# ---
# name: test_has_pii_content
  True
# ---
# name: test_has_pii_content.1
  False
# ---
# name: test_list_bucket
  <MagicMock name='mock.list_buckets()' id='140364591224416'>
# ---
